modules/apps/rate-limiting/keeper/packet.go

```diff
@@ -0,0 +1,268 @@
+package keeper
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	errorsmod "cosmossdk.io/errors"
+	sdkmath "cosmossdk.io/math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/cosmos/ibc-go/v10/modules/apps/rate-limiting/types"
+	transfertypes "github.com/cosmos/ibc-go/v10/modules/apps/transfer/types"
+	channeltypes "github.com/cosmos/ibc-go/v10/modules/core/04-channel/types"
+	channeltypesv2 "github.com/cosmos/ibc-go/v10/modules/core/04-channel/v2/types"
+)
+
+type RateLimitedPacketInfo struct {
+	ChannelID string
+	Denom     string
+	Amount    sdkmath.Int
+	Sender    string
+	Receiver  string
+}
+
+// CheckAcknowledementSucceeded unmarshals IBC Acknowledgements, and determines
+// whether the tx was successful
+func (k Keeper) CheckAcknowledementSucceeded(ctx sdk.Context, ack []byte) (success bool, err error) {
+	// Check if the ack is the IBC v2 universal error acknowledgement
+	if bytes.Equal(ack, channeltypesv2.ErrorAcknowledgement[:]) {
+		return false, nil
+	}
+
+	// Unmarshal the raw ack response
+	var acknowledgement channeltypes.Acknowledgement
+	if err := transfertypes.ModuleCdc.UnmarshalJSON(ack, &acknowledgement); err != nil {
+		return false, errorsmod.Wrapf(sdkerrors.ErrUnknownRequest, "cannot unmarshal ICS-20 transfer packet acknowledgement: %s", err.Error())
+	}
+
+	// The ack can come back as either AcknowledgementResult or AcknowledgementError
+	// If it comes back as AcknowledgementResult, the messages are encoded differently depending on the SDK version
+	switch response := acknowledgement.Response.(type) {
+	case *channeltypes.Acknowledgement_Result:
+		if len(response.Result) == 0 {
+			return false, errorsmod.Wrapf(channeltypes.ErrInvalidAcknowledgement, "acknowledgement result cannot be empty")
```

Comment:
Take a look and see if you can test these corner cases, would be good to have scenarios that cover them! Especially since our e2e tests are focusing on happy path